rfcs/20210301-computation-datatype/README.md

```diff
@@ -0,0 +1,178 @@
+user would be able to enable lower precision only for sub-graphs in
+the whole topology in order to control accuracy loss.
+
+It has to be noted that the implicit down-convert feature is planned
+to be exposed to end-user (FWK or toolkit users) as both a global
+setting and/or scoped to a sub-graph.
+
+# 2. Main considerations for the feature behavior
+
+Before making a proposal of API changes, let's first consider a few
+traits that will define the behavior of that feature.
+
+## Scoped vs global setting
+A global setting would be a library state, possibly controlled with an
+environment variable and/or a function call. The benefit is that it is
+easy to use and would require very little change to user
+code. However, a scoped setting is necessary if some users want to
+allow lower precision for selected portions of the graph.
+
+Proposal: use scoped setting, with a global (`thread_local`)
+default. This will allow the scoped use case the framework would
+expose, and also allow users to test implicit down-conversion without
+any framework change (with environment variable).
+
+## Allow only "compatible" down-conversion vs any data-type.
+
+Here compatible data-types are data-types that are typically used in
+the same way, and that are at least as precise. This excludes implicit
+conversion to/from any integer data-type, since integer data-types
+typically require quantization parameters that depend on the
+computation data-type. For floating point data-types
+- `bf16` and `f32` are compatible for both training and inference,
+- and `f16` and `f32` only for inference since `f32` and `f16` are not
+  typically interchangeable for training.
+
+Proposal: support only down-conversion to "compatible"
+data-types. This will allow to restrict the feature to only commonly
+used combinations and would make internal logic and validation
+simpler. Here we would return `invalid_argument` if users pass
+incompatible data-type for down-conversion hint/requirement, to help
+them debug incorrect logic.
+
+## Down-conversion policy: requirement vs hint.
+Here the question is if the user specifies a target type to
+down-convert to, should we treat it as a requirement (we can
+down-convert only to this type) or should it be a hint (we can
+down-convert to a type other than the one specified as long as it is
+"compatible").
+
+Proposal: support only hints, to allow library for optimal dispatching
+without user needing to tweak down-conversion policy for better
+performance. Also, this is more scalable as we introduce new
+data-types (e.g. if we were to introduce `tf32`).
+
+## Per tensor vs per primitive setting
+This would be the granularity at which the user would allow conversion
+to happen. Per tensor seems to have the most flexibility, however it
+is unlikely that users would need this level of flexibility. For the
+per-primitive setting, the implicit conversion will be assumed to
+happen only for input tensors, and not output.
```

Comment:
What data type will the output be? Will it be kept in the down converted data type or up converted back to fp32?

Reply:
Hi @jeyblu, when the inputs of a primitive are floating-point datatypes, oneDNN does all the accumulations in `fp32`, and could do a conversion after the computation if the user request it. For example, if a user creates a convolution with `src(f32)`,  `weights(f32)` and `dst(s8)`, the actual computation will happen in f32, then the operation output would be downconverted to `s8` before we write back to `dst`.

With this proposal, we would allow oneDNN implementations to implicitly convert `src` and `weights` to `f16` or `bf16`, while keeping the accumulation datatype as `f32`. So the computation would be
```
src(f32)     -> downconvert -+-> op(bf16, bf16, f32) -> postops -> final conversion to dst type.
                                       ^
weights(f32) -> downconvert -----------| 
```

Reply:
What if `dst` is specified as `any`?

Reply:
Oh my bad, now I understand the confusion in the RFC. `Implicit` here refers to the fact that the conversion transparent to the user (they do not know which down-conversion happened inside the primitive implementation). Here this sentence highlights that we would allow only down-conversion of activation/weights, but will strictly maintain accumulation and output to `f32` (datatype specified by user).

Regarding the case of a user passing `any`, this is the topic of the next section, and there are two options (specified in next section):
1. we don't expose down-conversion in queried memory descriptors. In that case, it is fully implicit mode, where all down-conversions happen internally, and only on the weights/activation.
2. we expose down-conversion in queried memory descriptors. In that case, it is semi-implicit in the sense that the implementation chooses to down-convert, but those are expose to user can incur extra reorder overhead if the down-converted memory cannot be propagated as is to the next layer.

In the next section, we propose to go with option 2, but in practice I expect our implementations to output f32 memory descriptors for activations (to minimize potential reorder overheads), and expose down-conversion only for weights.

Reply:
Thanks for the reply. Still have some questions. The user sets layout format for inputs and outputs. So there could be several cases to consider, e.g. all inputs and outputs are set to `any`, only some inputs and some outputs are set to `any`, all the inputs and outputs are set to a format other than `any`. What datatype will the output be if user sets output to `any`? What datatype will the output be if user sets output to a format other than `any`?

Reply:
Sure. For convolution, let's consider we have only src, weights and dst (so forward without bias). Then let's consider this code (mostly for naming):
```c++
auto conv_desc = convolution_forward::desc(prop_kind::forward,
        algorithm::convolution_direct, src_md, weights_md, dst_md,
        strides, padding_l, padding_r);
auto conv_pd
        = convolution_forward::primitive_desc(conv_desc, engine);
auto impl_src_md = conv_pd.src_desc();
auto impl_weights_md = conv_pd.weights_desc();
auto impl_dst_md = conv_pd.dst_desc();
```

So `src_md`, `weights_md` and `dst_md` are the descriptors passed by the user to create the convolution primitive.
And `impl_src_md`, `impl_weights_md` and `impl_dst_md` are the descriptors returned by the primitive implementation when queried.

To simplify, I will use `id` for implementation defined in the table below:
| descriptor | user tag | user type | impl tag | impl type ( downconvert disabled) | impl type ( downconvert enabled) |
|------------|----------|-----------|----------|-----------------------------------|----------------------------------|
| src        | any      | f32       | id       | f32                               | id                               |
|            | nhwc     | f32       | nhwc     | f32                               | f32                              |
| weights    | any      | f32       | id       | f32                               | id                               |
|            | oihw     | f32       | oihw     | f32                               | f32                              |
| dst        | any      | f32       | id       | f32                               | id                               |
|            | nhwc     | f32       | nhwc     | f32                               | f32                              |


As you can see, an implementation is allowed to return a different datatype in queried memory descriptor only if the user passed the `any` layout for that tensor.

Reply:
Thanks :) The example with the table is very helpful. You may want to consider including it in your documentation.

Reply:
Makes sense. I updated the RFC with that example. Thanks!